### PR TITLE
SDS filesystem client: additional methods, command line script for html report

### PR DESCRIPTION
**Pull request by megies**
Here's some more methods, mainly useful for QC checks on local SDS structure (latency, data availability, gaps/overlaps).

Works as is, but I should still add one or two simple tests..

### CODE REVIEW

File: obspy/clients/filesystem/sds.py [hunk 1]
```diff
@@ -108,7 +113,7 @@ def __init__(self, sds_root, sds_type="D", format="MSEED",
         self.fileborder_samples = fileborder_samples
 
     def get_waveforms(self, network, station, location, channel, starttime,
-                      endtime, merge=-1, sds_type=None):
+                      endtime, merge=-1, sds_type=None, **kwargs):
```
Comment (QuLogic) [inline obspy/clients/filesystem/sds.py]: Should be documented as (mostly) extra args for specific formats.

File: obspy/clients/filesystem/sds.py [hunk 2]
```diff
@@ -169,24 +224,274 @@ def get_waveforms(self, network, station, location, channel, starttime,
...
+        st.traces = [tr for tr in st
+                     if not (tr.stats.endtime < starttime or
+                             tr.stats.starttime > endtime)]
+
+        if not st:
+            return (0, 1)
+
+        total_duration = endtime - starttime
+        # sum up gaps in the middle
+        gaps = [gap[6] for gap in st.getGaps()]
```
Comment (QuLogic) [inline obspy/clients/filesystem/sds.py]: Hmm, was `getGaps` not renamed?

File: obspy/clients/filesystem/sds.py [hunk 3]
```diff
@@ -169,24 +224,274 @@ def get_waveforms(self, network, station, location, channel, starttime,
...
+                if st:
+                    break
+            time -= 24 * 3600
+
+        return max([tr.stats.endtime for tr in st])
+
+    def get_latency(self, network, station, location, channel,
+                    sds_type=None, stop_time=None):
+        """
+        Get time of last sample for given stream.
```
Comment (QuLogic) [inline obspy/clients/filesystem/sds.py]: It's not really the "time of", though...

File: obspy/clients/filesystem/sds.py [hunk 4]
```diff
@@ -169,24 +224,274 @@ def get_waveforms(self, network, station, location, channel, starttime,
...
+        :type location: str
+        :param location: Location code of requested data (e.g. "").
+        :type channel: str
+        :param channel: Channel code of requested data (e.g. "HHZ").
+        :type sds_type: str
+        :param sds_type: Override SDS data type identifier that was specified
+            during client initialization.
+        :type stop_time: :class:`~obspy.core.utcdatetime.UTCDateTime`
+        :param stop_time: Time at which the search for data is stopped and
+            `None` is returned. If not specified, stops at ``1950-01-01T00``.
```
Comment (QuLogic) [inline obspy/clients/filesystem/sds.py]: `None` should probably have two backticks?

File: obspy/clients/filesystem/sds.py [hunk 5]
```diff
@@ -169,24 +224,274 @@ def get_waveforms(self, network, station, location, channel, starttime,
...
```
Comment (QuLogic) [inline obspy/clients/filesystem/sds.py]: `None` should probably have two backticks? It does on the next line...

File: obspy/clients/filesystem/sds.py [hunk 6]
```diff
@@ -169,24 +224,274 @@ def get_waveforms(self, network, station, location, channel, starttime,
...
+        :param sds_type: Override SDS data type identifier that was specified
+            during client initialization.
+        :rtype: bool
+        """
+        sds_type = sds_type or self.sds_type
+        pattern = self.FMTSTR.format(
+            year="*", network=network, station=station, location=location,
+            channel=channel, doy=0, type=sds_type)
+        # can not insert wildcard for day-of-year above, so replace it now
+        pattern = pattern.rsplit(".", 1)[0] + "*"
```
Comment (QuLogic) [inline obspy/clients/filesystem/sds.py]: This assumes you're using the default pattern, doesn't it?

File: obspy/clients/filesystem/sds.py [hunk 7]
```diff
@@ -169,24 +224,274 @@ def get_waveforms(self, network, station, location, channel, starttime,
...
+        :rtype: list
+        :returns: List of (network, station, location, channel) 4-tuples of all
+            available streams in archive.
+        """
+        sds_type = sds_type or self.sds_type
+        pattern = self.FMTSTR.format(
+            year="*", network="*", station="*", location="*",
+            channel="*", doy=0, type=sds_type)
+        # can not insert wildcard for day-of-year above, so replace it now
+        pattern = pattern.rsplit(".", 1)[0] + "*"
```
Comment (QuLogic) [inline obspy/clients/filesystem/sds.py]: Same here.

File: obspy/clients/filesystem/sds.py [hunk 8]
```diff
@@ -169,24 +224,274 @@ def get_waveforms(self, network, station, location, channel, starttime,
...
+        pattern = os.path.join(self.sds_root, pattern)
+        all_files = glob.glob(pattern)
+        result = set()
+        for file_ in all_files:
+            network, station, channel, filename = file_.split("/")[-4:]
```
Comment (QuLogic) [inline obspy/clients/filesystem/sds.py]: This one too.

File: obspy/clients/filesystem/sds.py [hunk 9]
```diff
@@ -169,24 +224,274 @@ def get_waveforms(self, network, station, location, channel, starttime,
...
+        pattern = self.FMTSTR.format(
+            year="*", network="*", station="*", location="*",
+            channel="*", doy=0, type=sds_type)
+        # can not insert wildcard for day-of-year above, so replace it now
+        pattern = os.path.join(self.sds_root, pattern)
+        pattern = os.path.dirname(os.path.dirname(pattern))
+        all_folders = glob.glob(pattern)
+        result = set()
+        for file_ in all_folders:
+            network, station = file_.split("/")[-2:]
```
Comment (QuLogic) [inline obspy/clients/filesystem/sds.py]: And probably here, too.

**Comment by bonaime** [comment]
@megies: are you thinking of a method to import (and merge small) files from a directory to an SDS structure ?

File: obspy/clients/filesystem/sds.py (hunk 2 above)
Comment (megies) [inline obspy/clients/filesystem/sds.py]: Not *yet*.. I put a note in there https://github.com/obspy/obspy/issues/1039#issuecomment-150163012 already, so we do it before release..

File: obspy/clients/filesystem/sds.py (hunk 3 above)
Comment (megies) [inline obspy/clients/filesystem/sds.py]: Thanks.. copy/paste bug from `_get_endtime`..

File: obspy/clients/filesystem/sds.py (hunk 4 above)
Comment (megies) [inline obspy/clients/filesystem/sds.py]: Yep, thanks. We should at least have this right in new code. :+1:

File: obspy/clients/filesystem/sds.py (hunk 6 above)
Comment (megies) [inline obspy/clients/filesystem/sds.py]: Hmm, you're right, that's vulnerable.. I'll try to change this, replacing everything not in NSLC with `*`.

File: obspy/clients/filesystem/sds.py (hunk 7 above)
Comment (megies) [inline obspy/clients/filesystem/sds.py]: This was a bit tricky to generalize.. lots of regex magic in there now..

File: obspy/clients/filesystem/sds.py (hunk 8 above)
Comment (megies) [inline obspy/clients/filesystem/sds.py]: Same here.. lots of regex magic in there now..

File: obspy/clients/filesystem/sds.py (hunk 9 above)
Comment (megies) [inline obspy/clients/filesystem/sds.py]: Same here.. lots of regex magic in there now..

**Comment by megies** [comment]
> @megies: are you thinking of a method to import (and merge small) files from a directory to an SDS structure ?

I don't need that right now, so I'll not work on it right now, but that sounds like a useful addition.. feel free to give it a shot if you'd like. Although messing around with the actual data archive is something that I usually avoid to do too programmatically..

**Comment by megies** [comment]
This is what the basic SDS report html page looks like.. of course this is just for the dummy test SDS we use in obspy right now..

screenshot from 2015-12-08 11 33 17

File: obspy/clients/filesystem/sds.py (hunk 1 above)
Comment (megies) [inline obspy/clients/filesystem/sds.py]: I added a short note in the parameter description below, which should be enough..

**Comment by megies** [comment]
No idea why this seems to influence warnings of `read(..., headonly=True, startime=XXX, ...)`, because I put my call that is issuing the same warning that is failing to get shown in the Travis test within a `with warnings.catch_warnings()` context manager which is supposed to reset the original warning settings?? https://github.com/obspy/obspy/pull/1202/files#diff-93c710504a1d59384f7fc6ad859e2f06R297

**Comment by megies** [comment]
> context manager which is supposed to reset the original warning settings??

Obviously this is fixed in Py 3.x but not in 2.7..
https://bugs.python.org/issue4180#msg100525

**Comment by QuLogic** [comment]
Probably should squash those exploratory appveyor commits, and I think maybe change the colours a bit to match [Set2](http://colorbrewer2.org/) or similar. Also, are the colours in the table related to the colours in the figure? I think not, but I'm not sure...

**Comment by megies** [comment]
Yeah, of course already intended to squash all these mini commits at the end, before merging, don't worry. :+1: 
Just preferred to keep force-pushes to a minimum, otherwise nobody can follow whats going on here anymore until it's ready.

Color-wise.. this is mostly intended as a one-look quick assessment page for current state of health in local archive of a large number of stations. The harsh colors are kind of intentional (to highlight stations that are currently down or lagging a lot), and every color can be user set on command line.. so I think it's OK as is.

Colors in the figure are not related, simple reusing the `obspy-scan` output (KISS), just as another low-hanging-fruit level of detail, to get an idea of how gaps exactly behave over time in the current monitored time span..

**Comment by megies** [comment]
Colorbrewer Set1 looks better, you're right @QuLogic.. changed it. thanks!

**Comment by megies** [comment]
Will rebase/squash this..

**Comment by megies** [comment]
Merging..

**Pull request merged**